### DISCUSSION TITLE
How-to nest multiple submodules

### DISCUSSION THREAD

**Discussion by Sewtz** [pybind/pybind11#4027] [Q&A] [answered]
I am working on a larger Framework in C++ which is divided into several namespaces like 

```c++
namespace Framework::ModA{
   /* ModA stuff */
}

namespace Framework::ModB{
  /* ModB stuff */
}
```

Now I want to recreate such a structure in my python bindings. How can I achieve a structure like
```python
import Framework.ModA
import Framework.ModB
```

**Comment by jiwaszki** [comment, marked as answer]
Hi @Sewtz , I assume you want pure pybind solution. Here it is:
```cpp
#include <pybind11/pybind11.h>

namespace py = pybind11;

size_t free_function() {
      return 42;
}

namespace module_a {
      size_t func() {
            return 97;
      }
}

namespace module_b {
      size_t func() {
            return 20;
      }
}

PYBIND11_MODULE(mymodule, m)
{
      m.doc() = "This is top module - mymodule.";
      m.def("free_function", &free_function);

      auto m_a = m.def_submodule("module_a", "This is A.");
      m_a.def("func", &module_a::func);

      auto m_b = m.def_submodule("module_b", "This is B.");
      m_b.def("func", &module_b::func);
}
```
You can either go with this pure pybind code and register submodules according to names of your namespaces (you can chain them as well) or register modules and re-arrange them with init files.

So this is it, for example you can look up submodules in terminal using `help`:
```
Help on module mymodule:

NAME
    mymodule - This is top module - mymodule.

SUBMODULES
    module_a  <-- A is here
    module_b  <-- B is here

FUNCTIONS
    free_function(...) method of builtins.PyCapsule instance
        free_function() -> int
```

**Reply by Sewtz** [reply]
Thank you

That's what I needed.

**Comment by The9Cat** [comment]
@jiwaszki I hope you don't mind that I'm asking a question about your example.   There is something that has puzzled me for a bit.

The line
```
auto m_a = m.def_submodule("module_a", "This is A.");
```
defines a docstring for the submodule.  But running `help` as in the example above does not print the docstring, it just lists the module by name.  And it's not printed in `help(mymodule.module_a)` either.  What is this submodule docstring for?
